### PR TITLE
Bug 1412824 Added context to egress router section

### PR DESCRIPTION
**Pull request by bfallonf**
As per: https://bugzilla.redhat.com/show_bug.cgi?id=1412824

I'll ask for an ack in the BZ.

cc: @sferich888

### CODE REVIEW

File: admin_guide/managing_pods.adoc [hunk 1]
```diff
@@ -146,16 +146,17 @@ endif::openshift-enterprise,openshift-origin[]
 
 [[admin-guide-controlling-egress-traffic]]
 == Controlling Egress Traffic
-As an {product-title} cluster administrator, you can control egress traffic
-using either a xref:admin-guide-deploying-an-egress-router-service[router] or
-xref:admin-guide-limit-pod-access-egress[firewall] methods.
+As an {product-title} cluster administrator, you can control egress traffic in two different ways:
+
+* Using an egress xref:admin-guide-deploying-an-egress-router-service[router], so that an external source treats traffic as though it were coming from a known source.
+* Using an egress xref:admin-guide-limit-pod-access-egress[firewall], where traffic is accepted, enforced, and defined.
```
Comment (sferich888) [inline admin_guide/managing_pods.adoc]: I think this is trying to say, what https://docs.openshift.com/container-platform/3.3/admin_guide/managing_pods.html#admin-guide-limit-pod-access-egress says, but it does not quite align. 

In short `accepted, enforced, and defined` (are not how I would phrase this). 

In stead I would say: 

`Using an egress xref:admin-guide-limit-pod-access-egress[firewall] allows a cluster administrator to enforce acceptable (outbound) traffic policies. These policies can be defined in such a way that, acceptable endpoints or ip ranges (subnets) are the only acceptable targets for the dynamic endpoints (pods with in OpenShift) to talk to.`

File: admin_guide/managing_pods.adoc [hunk 2]
```diff
@@ -146,16 +146,17 @@ endif::openshift-enterprise,openshift-origin[]
 
 [[admin-guide-controlling-egress-traffic]]
 == Controlling Egress Traffic
-As an {product-title} cluster administrator, you can control egress traffic
-using either a xref:admin-guide-deploying-an-egress-router-service[router] or
-xref:admin-guide-limit-pod-access-egress[firewall] methods.
+As an {product-title} cluster administrator, you can control egress traffic in two different ways:
+
+* Using an egress xref:admin-guide-deploying-an-egress-router-service[router], so that an external source treats traffic as though it were coming from a known source.
+* Using an egress xref:admin-guide-limit-pod-access-egress[firewall], where a cluster administrator can enforce acceptable (outbound) traffic policies, so that specific endpoints or IP ranges (subnets) are the only acceptable targets for the dynamic endpoints (pods within {product-title}) to talk to.
```
Comment (sferich888) [inline admin_guide/managing_pods.adoc]: `Using an egress xref:admin-guide-limit-pod-access-egress[firewall], where a cluster administrator` does not read right. Should this say `Using an egress xref:admin-guide-limit-pod-access-egress[firewall], allows a cluster administrator`?

File: admin_guide/managing_pods.adoc [hunk 3]
```diff
@@ -146,16 +146,126 @@ endif::openshift-enterprise,openshift-origin[]
...
+
+xref:admin-guide-limit-pod-access-egress[Firewall]::
+Using an egress firewall allows you to enforce the acceptable outbound traffic
+policies, so that specific endpoints or IP ranges (subnets) are the only
+acceptable targets for the dynamic endpoints (pods within {product-title}) to
+talk to.
+
+xref:admin-guide-limit-pod-access-egress-router[Router]::
+Using an egress router allows you to create identifiable services to send
+traffic to a specific source, ensuring an external source treats traffic as
```
Comment (danwinship) [inline admin_guide/managing_pods.adoc]: "destination" not "source" (both times)

File: admin_guide/managing_pods.adoc [hunk 4]
```diff
@@ -146,16 +146,126 @@ endif::openshift-enterprise,openshift-origin[]
...
+
+* If you merge two projects together (via `oadm pod-network join-projects`),
+then you cannot use egress policy in _any_ of the joined projects.
+
+* No project may have more than one egress policy object.
+
+Violating any of these restrictions results in broken egress policy for the
+project, and may cause all external network traffic to be dropped.
+
+Use the `oc` command or the REST API to configure pod access limits. You can use
```
Comment (danwinship) [inline admin_guide/managing_pods.adoc]: Probably it's best to get rid of the "pod access limits" language altogether? "to configure egress policy" (and then again below)

File: admin_guide/managing_pods.adoc [hunk 5]
```diff
@@ -146,16 +146,126 @@ endif::openshift-enterprise,openshift-origin[]
...
+
+. Navigate to the project you want to affect.
+
+. Create a JSON file for the pod limit policy:
++
+----
+# oc create -f <policy>.json
+----
+
+. Configure the JSON file with policy details. For example:
```
Comment (danwinship) [inline admin_guide/managing_pods.adoc]: This is backwards... you have to create the JSON file first, then call "oc create -f <policy>.json" (which reads *from* the file, to create the object in OpenShift)

File: admin_guide/managing_pods.adoc [hunk 6]
```diff
@@ -219,61 +331,45 @@ endif::openshift-origin[]
...
+<3> Same value as the default gateway used by the node itself.
+<4>  Connections to the pod are redirected to 203.0.113.25, with a source IP
+address of 192.168.12.99
+<5> The pod will only be deployed to nodes with the label site `springfield-1`.
+====
++
 [NOTE]
 ====
 Preserve the the quotation marks around `"true"`. Omitting them will result in
 errors.
```
Comment (danwinship) [inline admin_guide/managing_pods.adoc]: move this into note <1> now?

File: admin_guide/managing_pods.adoc [hunk 7]
```diff
@@ -219,61 +331,45 @@ endif::openshift-origin[]
...
-The `pod.network.openshift.io/assign-macvlan annotation` creates a Macvlan
+<1> The `pod.network.openshift.io/assign-macvlan annotation` creates a Macvlan
 network interface on the primary network interface, and then moves it into the
 pod's network name space before starting the *egress-router* container.
-
+<2> IP address on the node subnet reserved by the cluster administrator for use by this pod.
+<3> Same value as the default gateway used by the node itself.
+<4>  Connections to the pod are redirected to 203.0.113.25, with a source IP
+address of 192.168.12.99
+<5> The pod will only be deployed to nodes with the label site `springfield-1`.
```
Comment (danwinship) [inline admin_guide/managing_pods.adoc]: "site" is part of the label. It should say

    ...with the label `site=springfield-1`.

File: admin_guide/managing_pods.adoc [hunk 8]
```diff
@@ -146,16 +146,126 @@ endif::openshift-enterprise,openshift-origin[]
...
+Using an egress router allows you to create identifiable services to send
+traffic to a specific source, ensuring an external source treats traffic as
+though it were coming from a known source. This helps with security, because it
+allows you to secure an external database so that only specific pods in a
+namespace can talk to a service (the egress router), which proxies the traffic
+to your database.
+
+
+[[admin-guide-limit-pod-access-egress]]
+=== Limiting Pod Access with Egress Firewall
```
Comment (danwinship) [inline admin_guide/managing_pods.adoc]: These titles (this one and "Limiting Pod Access with an Egress Router") should probably be changed too, to make it clearer that these are two fairly different things. Maybe "Limiting Access to External Resources with an Egress Firewall" and "Allowing External Resources to Recognize Pod Traffic with an Egress Router" or something

Comment (bfallonf) [inline admin_guide/managing_pods.adoc]: I like it. It helps with Eric's initial BZ request about making the two procedures more different.

**Comment by bfallonf** [comment]
Thanks, @danwinship . I've made some updates as per your comments, added the answers you gave in the BZ, and added the section on egress router failover. Can I ask for a final ack or if there are any more issues? Thanks in advance.

cc: @sferich888 as the original BZ reporter, let me know if there's anything else you'd like to add.

File: admin_guide/managing_pods.adoc [hunk 9]
```diff
@@ -146,16 +146,126 @@ endif::openshift-enterprise,openshift-origin[]
...
+xref:admin-guide-limit-pod-access-egress[Firewall]::
+Using an egress firewall allows you to enforce the acceptable outbound traffic
+policies, so that specific endpoints or IP ranges (subnets) are the only
+acceptable targets for the dynamic endpoints (pods within {product-title}) to
+talk to.
+
+xref:admin-guide-limit-pod-access-egress-router[Router]::
+Using an egress router allows you to create identifiable services to send
+traffic to a specific destination, ensuring an external source treats traffic as
+though it were coming from a known destination. This helps with security,
```
Comment (danwinship) [inline admin_guide/managing_pods.adoc]: ...traffic to a specific destination, ensuring an external **destination** treats traffic as though it were coming from a known **source**.

File: admin_guide/managing_pods.adoc [hunk 10]
```diff
@@ -146,16 +146,126 @@ endif::openshift-enterprise,openshift-origin[]
...
+Use the `oc` command or the REST API to configure egress policy. You can use
+`oc [create|replace|delete]` to manipulate `EgressNetworkPolicy` objects. The
+*_api/swagger-spec/oapi-v1.json_* file has API-level details on how the objects
+actually work.
+
+To configure egress policy:
+
+. Navigate to the project you want to affect.
+
+. Configure the JSON file with policy details. For example:
```
Comment (danwinship) [inline admin_guide/managing_pods.adoc]: "What JSON file?"

Should say something like: "Create a JSON file with policy details. For example:"

and then the next bullet point (which currently says "Create a JSON file...") should be something like "Apply that policy to the project" or "Create an EgressNetworkPolicy object from the JSON file".

File: admin_guide/managing_pods.adoc [hunk 11]
```diff
@@ -219,32 +331,63 @@ endif::openshift-origin[]
...
+network interface on the primary network interface, and then moves it into the
+pod's network name space before starting the *egress-router* container. Preserve the the quotation marks around `"true"`. Omitting them will result in
+errors.
+<2> An IP address from the physical network that the node itself is on and is
+reserved by the cluster administrator for use by this pod.
+<3> Same value as the default gateway used by the node itself.
+<4> The external server that to direct traffic to. Using this example, connections
+to the pod are redirected to 203.0.113.25, with a source IP address of
+192.168.12.99.
+<5> The pod will only be deployed to nodes with the label `springfield-1`.
```
Comment (danwinship) [inline admin_guide/managing_pods.adoc]: \`site=springfield-1\`, not just \`springfield-1\`. This line should be exactly the following text:

    <5> The pod will only be deployed to nodes with the label `site=springfield-1`.

File: admin_guide/managing_pods.adoc [hunk 12]
```diff
@@ -219,32 +331,63 @@ endif::openshift-origin[]
...
 ====
 
-The `pod.network.openshift.io/assign-macvlan annotation` creates a Macvlan
-network interface on the primary network interface, and then moves it into the
-pod's network name space before starting the *egress-router* container.
+. Create the pod using the definition:
++
+----
+$ oc create -f <pod_name>
```
Comment (danwinship) [inline admin_guide/managing_pods.adoc]: You have to pass the name of the JSON file created in the previous step, not the pod name. (But in the "oc get" example below, using pod_name is correct.)

It looks like just saying "<pod_name>.json" might be consistent with other examples

File: admin_guide/managing_pods.adoc [hunk 13]
```diff
@@ -268,129 +411,62 @@ If only some of the nodes in your cluster are capable of claiming the specified
...
-
-To configure pod access limits:
-
-. Navigate to the project you want to affect.
-. Create a JSON file for the pod limit policy:
+. Create the pod using the definition:
 +
 ----
-# oc create -f <policy>.json
+$ oc create -f <replication_controller>
```
Comment (danwinship) [inline admin_guide/managing_pods.adoc]: add ".json"

**Review by danwinship — COMMENTED** [review]
Looks good. A few things that aren't quite right or that I didn't notice before:

**Comment by bfallonf** [comment]
Thanks @danwinship Let me know if there's anything else.

In the meantime, @ahardin-rh @bmcelvee @adellape peer review please?

File: admin_guide/managing_pods.adoc [hunk 14]
```diff
@@ -146,16 +146,138 @@ endif::openshift-enterprise,openshift-origin[]
...
+internal hosts (outside the cluster).
++
+Or,
+- A pod cannot reach specified internal subnets/hosts that it should have no
+reason to contact.
+
+You can configure projects to have different egress policies. For example,
+allowing `<project A>` access to a specified IP range, but denying the same
+access to `<project B>`. Or restrict application developers from updating from
+(python) pip mirrors, and forcing updates to only come from desired sources.
```
Comment (ahardin-rh) [inline admin_guide/managing_pods.adoc]: s/python/Python

File: admin_guide/managing_pods.adoc [hunk 15]
```diff
@@ -219,32 +343,63 @@ endif::openshift-origin[]
...
-this pod.
-<2> Same value as the default gateway used by the node itself.
-<3>  Connections to the pod are redirected to 203.0.113.25, with a source IP
-address of 192.168.12.99
-<4> The pod will only be deployed to nodes with the label site *springfield-1*.
+    site: springfield-1 <5>
+----
+<1> The `pod.network.openshift.io/assign-macvlan annotation` creates a Macvlan
+network interface on the primary network interface, and then moves it into the
+pod's network name space before starting the *egress-router* container. Preserve the the quotation marks around `"true"`. Omitting them will result in
```
Comment (ahardin-rh) [inline admin_guide/managing_pods.adoc]: Maybe wrap all of <1>

**Comment by ahardin-rh** [comment]
@bfallonf just minor nits from me :ok_hand:

**Comment by bfallonf** [comment]
Thanks @ahardin-rh . I'll merge.

**Comment by bfallonf** [comment]
[rev_history]
|xref:../admin_guide/managing_pods.adoc#admin-guide-manage-pods[Managing Pods]
|Added information and arranged the  xref:../admin_guide/managing_pods.adoc#admin-guide-controlling-egress-traffic[Controlling Egress Traffic] section.
%

**Pull request merged**